crates/rpc/rpc-eth-api/src/helpers/block.rs

```diff
@@ -233,18 +233,41 @@ pub trait LoadBlock: LoadPendingBlock + SpawnBlocking {
                 };
             }
 
+            // We retrieve the block hash for the given block id via the provider
             let block_hash = match LoadPendingBlock::provider(self)
                 .block_hash_for_id(block_id)
                 .map_err(Self::Error::from_eth_err)?
             {
                 Some(block_hash) => block_hash,
+                // If the block hash is not found, we return `None` directly
                 None => return Ok(None),
             };
 
-            self.cache()
-                .get_sealed_block_with_senders(block_hash)
-                .await
-                .map_err(Self::Error::from_eth_err)
+            // Initialize the maximum number of retries for handling reorg cases.
+            //
+            // A reorg may cause the latest block to be temporarily absent from the cache.
+            //
+            // By retrying once, we give the system a chance to update the cache with
+            // the new latest block information after a reorganization.
+            let max_retries = 1;
+
+            // Attempt to fetch the block from cache
+            for _ in 0..=max_retries {
+                match self.cache().get_sealed_block_with_senders(block_hash).await {
+                    // If a block is found in the cache, return it
+                    Ok(Some(block)) => return Ok(Some(block)),
+                    // If no block is found and the `block_id` refers to the latest block,
+                    // we retry the fetch, as this may indicate a reorg scenario
+                    Ok(None) if block_id.is_latest() => continue,
```

Comment:
here we'd need to update the hash again and check if it's different

Reply:
I just fixed this case by recalculating, sorry for this, was stupid but the scope here was not 100% clear to me at the beginning :)

Reply:
Let me know if this is ok to you or if you feel we can improve:
- The loop should just do a single iteration if not latest (so should be as it was before for this case).
- This `Ok(None) if block_id.is_latest()` is never reached if not latest so this should be ok too.